cloud_functions/gcs_event_based_ingest/README.md

```diff
@@ -0,0 +1,248 @@
+[supported default hive partitioning](https://cloud.google.com/bigquery/docs/hive-partitioned-loads-gcs#supported_data_layouts)
+you can specify this in the [`hivePartitioningOptions`](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#hivepartitioningoptions)
+key of `load.json` for that table.
+
+Any non-trivial incremental loading to partitions should usually use the
+Transformation SQL to define the `INSERT / MERGE / UPDATE / DELETE` logic into
+the target BQ table as these DML semantics are much more flexible thant the load
+job write dispositions.
+Furthermore, using external query has the added benefit of circumventing the 
+per load job bytes limits (default 15 TB) and commiting large partitions
+atomically.
+
+## Handling Incremental Loads
+This solution introduces the concept of `batch_id` which uniquely identifies 
+a batch of data committed by an upstream system that needs to be picked up as an
+incremental load. You can again set the load job or external query configuration
+at any parent folders `_config` prefix. This allows you dictate
+"for this table any new batch should `WRITE_TRUNCATE` it's parent partition/table"
+or "for that table any new batch should `WRITE_APPEND` to it's parent partition/table".
+
+## Monitoring
+Monitoring what data has been loaded by this solution should be done with the
+BigQuery [`INFORMATION_SCHEMA` jobs metadata](https://cloud.google.com/bigquery/docs/information-schema-jobs)
+If more granular data is needed about a particular job id 
+
+### Job Naming Convention
+All load or external query jobs will have a job id witha  prefix following this convention:
+```python3
+job_id_prefix=f"gcf-ingest-{dest_table_ref.dataset_id}-{dest_table_ref.table_id}-{1}-of-{1}-"
+```
+
+### Job Labels
+All load or external query jobs are labelled with functional component and cloud function name.
+```python3
+DEFAULT_JOB_LABELS = {
+    "component": "event-based-gcs-ingest",
+    "cloud-function-name": getenv("FUNCTION_NAME"),
+    "gcs-prefix": gs://bucket/prefix/for/this/ingest,
+}
+```
+If the destination regex matches a batch group, there will be a `batch-id` label.
+
+### Example INFROMATION SCHEMA Query
+```sql
+SELECT
+   job_id,
+   job_type,
+   start_time,
+   end_time,
+   query,
+   total_bytes_processed,
+   total_slot_ms,
+   destination_table
+   state,
+   (SELECT value FROM UNNEST(labels) WHERE key = "component") as component,
+   (SELECT value FROM UNNEST(labels) WHERE key = "cloud-function-name") as cloud_function_name,
+   (SELECT value FROM UNNEST(labels) WHERE key = "batch-id") as batch_id,
+FROM
+   `region-us`.INFORMATION_SCHEMA.JOBS_BY_PROJECT
+WHERE
+   (SELECT value FROM UNNEST(labels) WHERE key = "component") = "event-based-gcs-ingest"
+```
+
+## Triggers
+
+### Pub/Sub Storage Notifications `_SUCCESS`
+1. Trigger on `_SUCCESS` File to load all other files in that directory.
+1. Trigger on non-`_SUCCESS` File will no-op
```

Comment:
I'm curious if there's an alternative way to implement this. For instance I see for incrementals we include batch id. But alternatively these semantics could be specified as:

Batch load if a success file is specified, otherwise load files as they land into GCS. If the user wants transactional semantics, they can use the success file method. The WRITE_DISPOSITION can be separately defined in the config such that either method could technically be incremental.

Reply:
Interesting proposal.  
tl;dr this should be addressed w/ another tool and if we want to support it in this function it should be a separate PR IMHO.

I had a similar thought but decided not to implement it this way. Do you have a customer with actual requirements like this? I'm open to thinking about it more but specific requirements would be helpful.
e.g. 
- freshness requirements (aka why not use scheduled loads every 2-5 mins? could even use this solution and drop a success file and write truncate the partition every few mins / or use transnational external query for more specificity)
- query performance requirements (aka why not use external table?)
- duplicate loading of same file tolerance 
- how many files per day / hour
- "files not loaded sequentially" tolerance

not sure what kind of queries don't care about the above things except speculative results which could usually be served w/ external table.

IMHO this use cases of "load each file as it lands" are 90%  of the time either better served with external query the data directly OR pivoting publishers to streaming inserts.

I've originally avoided this because it will create an inefficient number of load jobs (resulting in inefficient number of storage sets IIUC which may damage performance of queries until some compaction of storage sets occurs) and I've been focused on scalability.
It will become much more likely to hit max load jobs per project / table https://cloud.google.com/bigquery/quotas#load_jobs.
This also makes the system harder to monitor / reason about.

Also this would complicate handling duplicate OCN pubsub messages as we'd have to use a more sophisticated de-duplication strategy than the one currently implemented (based on a single lock file). This will likely require a transactional database (cloud SQL, firestore, memorystore, bigtable). In my experience customers doing BQ migrations don't have a lot of GCP services approved by their infosec teams yet making additional complexity in the architecture a blocker to many orgs, not to mention adding cost / maintenance overhead of this metadata db to an otherwise cheap / managed solution.